core/solver/trs.cpp

```diff
@@ -0,0 +1,112 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2019, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/trs.hpp>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/solver/trs_kernels.hpp"
+
+namespace gko {
+namespace solver {
+
+
+namespace trs {
+
+
+GKO_REGISTER_OPERATION(generate, trs::generate);
+GKO_REGISTER_OPERATION(solve, trs::solve);
+
+
+}  // namespace trs
+
+
+template <typename ValueType, typename IndexType>
+void Trs<ValueType, IndexType>::generate(const LinOp *system_matrix,
+                                         const LinOp *b)
+{
+    using CsrMatrix = matrix::Csr<ValueType, IndexType>;
+    using Vector = matrix::Dense<ValueType>;
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+    // This is needed because it does not make sense to call the copy and
+    // convert if the existing matrix (if not CSR) is empty.
+    if (dynamic_cast<const CsrMatrix *>(system_matrix) == nullptr) {
```

Comment:
Yes, I do allow for empty creation with CSR matrices. This is necessary in case you just want to initialize a solver without passing in a matrix and want to copy it from another solver later.

Reply:
The solver kernels currently can solve only with a CSR matrix. So, initializing the solver with a CSR matrix does not require any conversion to CSR because the `copy_and_convert_to` does not need to be called. So, you can pass in a empty CSR matrix to initialize the solver and copy it from another solver if necessary. See [copy test](https://github.com/ginkgo-project/ginkgo/blob/trs_solver/reference/test/solver/trs.cpp#L95) 

But if you pass in an empty dense matrix, the generate step has to first convert it to CSR. This does not make sense because it does not make sense to convert an empty Dense matrix.

When the user passes in an empty dense matrix, you could technically not take that and initialize with an empty CSR matrix but that would be behaviour hidden from the user, which is not good I think.

Reply:
I thought it should work to convert an empty Dense matrix to an empty CSR matrix (I have not tested it though).
The current implementation restricts the user a bit too much because we force him/her to use a CSR matrix in the empty case, while any other format (that is convertable to CSR) can be used if the matrix is not empty. This is inconsistent and also a totally unnecessary restriction in my opinion.

If you want to handle the empty case separately (or if an empty matrix cannot be converted, which should be tested), you can add a test in the beginning to see if the given matrix is empty, and if it is, create an empty CSR matrix without calling a conversion.

Reply:
Yes, unfortunately, I dont think the conversion works on empty matrices. They just seg-fault which is definitely not good. I think it should be possible to give out some kind of an error.